### PR TITLE
Proof of concept

### PR DESCRIPTION
**Pull request by jorgenbuilder**
Attempt at making a dumb end-to-end PoC: https://github.com/metascore/metascore/issues/12

These commits include:

- Heartbeat
- Game can self registration
- Metascore can score loop (query -> percentile -> metascore)

My first priority at the moment is to prove out the difficult technical bits. Additions and thoughts are welcome!

### CODE REVIEW

File: src/backend/main.mo [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+
+    ////////////
+    // State //
+    //////////
+
+    stable var owners : [Principal] = [];
+    stable var gameCanDir : [Types.GameCanRecord] = [];
+    // I'm sure theres a better data model to use. Tries?
+    stable var playerScores : [(Types.Player, Types.GameCanRecord, Types.Score)] = [];
```
Comment (di-wu) [inline src/backend/main.mo:24]: I'll let this slide for, since this is a POC....

**Review by di-wu — COMMENTED** [review]
Here are my 2¢ (a.k.a. thoughts) on this matter.
It might not be very organised, or in any specific order.

Feel free to comment if you have any questions.

---

First I imagine one central metascore canister that is used in the frontend to display information.
This canister is responsible for orchestrating this whole this.

What does this mean, 'orchestrating'? I will omit any countermeasures that has to do with sybil-attacks for the time being.

1. A canister can register **itself** as a game canister by calling an endpoint on the metascore canister. This would contain information such as the name, creators, etc. This would also prevent the registration of too many canisters since the canister itself need to exist first.
2. The canister, that just called this endpoint, also need to implement some kind of `scoring` endpoint so the metascore canister can call this on a certain interval. This information then is stored (\*).

---

Additional/Other options:

- (**\***) Let games host their own score canisters, the metascore canister just redirects the traffic.
- We create a new canister to store the scores of (verified/big) games to not overload the metascore canister.

File: src/backend/main.mo (hunk 1 above)
Comment (jorgenbuilder) [reply src/backend/main.mo:24]: If you are interested in writing the data model that would be great 😆

**Comment by jorgenbuilder** [comment]
> 1. A canister can register **itself** as a game canister by calling an endpoint on the metascore canister. This would contain information such as the name, creators, etc. This would also prevent the registration of too many canisters since the canister itself need to exist first.

Exactly how I picture this working. To this PoC we would just add a frontend that allows the user plug in their Game Can's principal.

> 2. The canister, that just called this endpoint, also need to implement some kind of scoring endpoint so the metascore canister can call this on a certain interval. This information then is stored (*).

Also exactly how I picture this working. The scoring endpoint `metascoreDump` in this PoC is essentially that.

File: src/backend/main.mo (hunk 1 above)
Comment (di-wu) [reply src/backend/main.mo:24]: Coming up. 🔜

Comment (di-wu) [reply src/backend/main.mo:24]: We first have to make the data model of course. 😊

**Comment by di-wu** [comment]
The following commit https://github.com/metascore/metascore/commit/8f92c10b8a7e28131216b75928693c1d5b96ef5f contains another small POC that covers what I said before.

**Comment by jorgenbuilder** [comment]
@di-wu I think I missed some nuance of your first point. Game Canister registering itself in it's canister code is a cool idea, would love to see that fleshed out!

**Comment by jorgenbuilder** [comment]
Quint's is better.

**Pull request closed without merging**